### PR TITLE
Initial public draft for TotW 65

### PR DESCRIPTION
**Pull request by tituswinters**

### CODE REVIEW

File: _posts/2017-11-16-totw-65.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+that object into the container. Avoiding these copies is more efficient for
+almost all objects, and makes it easier to store move-only objects (such as
+`std::unique_ptr`) in standard containers.
+
+## The Old Way and the New Way
+
+Let’s look at a simple example using vectors to contrast the two styles. The
+first example uses pre-C++11 code:
+
+```
```
Comment (manshreck) [inline _posts/2017-11-16-totw-65.md]: \```c++

(here and throughout)

File: _posts/2017-11-16-totw-65.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  …
+};
+
+void addFoo() {
+  std::vector<Foo> v1;
+  v1.push_back(Foo(1, 2));
+}
+```
+
+Using the older `push_back()` method, two Foo objects are constructed: the
```
Comment (manshreck) [inline _posts/2017-11-16-totw-65.md]: \`Foo`

File: _posts/2017-11-16-totw-65.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+}
+```
+
+## Using Emplace Methods for Move-Only Operations
+
+So far, we’ve looked at cases where emplace methods improve performance, but
+they also make previously impossible code feasible, such as storing move-only
+types like `std::unique_ptr` within containers. Consider this snippet:
+
+    std::vector<std::unique_ptr<Foo>> v1;
```
Comment (manshreck) [inline _posts/2017-11-16-totw-65.md]: Don't use the indent version for code snippets. use The \```c++ version instead.

(Reason is below.)

File: _posts/2017-11-16-totw-65.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+How would you insert values into this vector? One way would be to use
+`push_back()` and construct the value directly within its argument:
+
+    v1.push_back(std::unique_ptr<Foo>(new Foo(1, 2)));
+
+This syntax works, but can be a bit unwieldy. Unfortunately, the traditional way
+of getting around this confusion is fraught with complexity:
+
+    Foo *f2 = new Foo(1, 2);
```
Comment (manshreck) [inline _posts/2017-11-16-totw-65.md]: Needs to be within \``` (because the * gets rendered as italics)

File: _posts/2017-11-16-totw-65.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    v1.push_back(std::unique_ptr<Foo>(f2));
+
+This code compiles, but it leaves ownership of the raw pointer unclear until the
+insertion. What’s worse, the vector now owns the object, but f2 still remains
```
Comment (manshreck) [inline _posts/2017-11-16-totw-65.md]: \`f2`

File: _posts/2017-11-16-totw-65.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/65
+type: markdown
+order: "065"
+---
+
+Originally posted as totw/65 on 2013-12-12
+
+By Hyrum Wright (hyrum@hyrumwright.org)
```
Comment (manshreck) [inline _posts/2017-11-16-totw-65.md]: Enclose email in mailto link:

[(hyrum@hyrumwright.org)](mailto:hyrum@hyrumwright.org)

(I know the mailto: is not as often used anymore, but the styling is also nice to have.)

**Comment by tituswinters** [comment]
Done, PTAL

File: _posts/2017-11-16-totw-65.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+}
+```
+
+## Using Emplace Methods for Move-Only Operations
+
+So far, we’ve looked at cases where emplace methods improve performance, but
+they also make previously impossible code feasible, such as storing move-only
+types like `std::unique_ptr` within containers. Consider this snippet:
+
+```c++
```
Comment (manshreck) [inline _posts/2017-11-16-totw-65.md]: FYI: nit, but when we add \```c++, we can eliminate the indentation. (That's one of the reasons I'd like to get rid of it as well.)

Comment (tituswinters) [reply _posts/2017-11-16-totw-65.md]: Done.

**Pull request merged**